pkg/cli/image/mirror/file.go

```diff
@@ -0,0 +1,441 @@
+	fi, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return distribution.Descriptor{}, distribution.ErrBlobUnknown
+		}
+		return distribution.Descriptor{}, err
+	}
+	if fi.IsDir() {
+		return distribution.Descriptor{}, fmt.Errorf("not a file")
+	}
+	return distribution.Descriptor{
+		Digest: dgst,
+		Size:   fi.Size(),
+	}, nil
+}
+
+func (s *fileBlobStore) Delete(ctx context.Context, dgst godigest.Digest) error {
+	return fmt.Errorf("unimplemented")
+}
+
+func (s *fileBlobStore) Get(ctx context.Context, dgst godigest.Digest) ([]byte, error) {
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "blobs", dgst.String())
+	return ioutil.ReadFile(path)
+}
+
+func (s *fileBlobStore) Open(ctx context.Context, dgst godigest.Digest) (distribution.ReadSeekCloser, error) {
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "blobs", dgst.String())
+	return os.Open(path)
+}
+
+func (s *fileBlobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, r *http.Request, dgst godigest.Digest) error {
+	return fmt.Errorf("unimplemented")
+}
+
+func (s *fileBlobStore) Put(ctx context.Context, mediaType string, payload []byte) (distribution.Descriptor, error) {
+	dgst := godigest.FromBytes(payload)
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "blobs", dgst.String())
+	if err := atomicWrite(path, payload); err != nil {
+		return distribution.Descriptor{}, err
+	}
+	return distribution.Descriptor{MediaType: mediaType, Size: int64(len(payload)), Digest: dgst}, nil
+}
+
+func (s *fileBlobStore) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
+	var opts distribution.CreateOptions
+	for _, option := range options {
+		err := option.Apply(&opts)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if opts.Mount.Stat == nil || len(opts.Mount.Stat.Digest) == 0 {
+		return nil, fmt.Errorf("file target blob store requires blobs to have mount stats that include a digest")
+	}
+	d := opts.Mount.Stat.Digest
+
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "blobs", d.String())
+
+	if opts.Mount.ShouldMount {
+		repoPath := repoPathForName(opts.Mount.From.Name())
+		sourcePath := filepath.Join(s.r.basePath, "v2", repoPath, "blobs", d.String())
+		klog.V(5).Infof("Try mount from %s to %s", sourcePath, path)
+		if fi, err := os.Stat(sourcePath); err == nil && !fi.IsDir() {
+			if err := atomicLink(path, sourcePath); err != nil {
+				return nil, err
+			}
+			return nil, distribution.ErrBlobMounted{
+				From:       opts.Mount.From,
+				Descriptor: distribution.Descriptor{Digest: d, Size: fi.Size()},
+			}
+		}
+	}
+	return s.r.newWriter(path, d.String(), opts.Mount.Stat.Size), nil
+}
+
+func (s *fileBlobStore) Resume(ctx context.Context, id string) (distribution.BlobWriter, error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// fileWriter attempts to upload parts to S3 in a buffered fashion where the last
```

Comment:
doesn't seem like this is doing anything related to s3.

Reply:
copy-pasta